test/cloudnative/agentversion_set.go

```diff
@@ -0,0 +1,183 @@
+	first := versions[0]
+	second := versions[len(versions)-1]
+
+	firstVersion, err := version.ExtractSemanticVersion(first)
+	require.NoError(t, err)
+	secondVersion, err := version.ExtractSemanticVersion(second)
+	require.NoError(t, err)
+
+	compare := version.CompareSemanticVersions(firstVersion, secondVersion)
+	if compare > 0 {
+		old = secondVersion
+		new = firstVersion
+	} else if compare < 0 {
+		old = firstVersion
+		new = secondVersion
+	}
+	return old, new
+}
+
+func updateDynakube(semanticVersion version.SemanticVersion) features.Func {
+	return func(ctx context.Context, t *testing.T, environmentConfig *envconf.Config) context.Context {
+		resources := environmentConfig.Client().Resources()
+		var dk dynatracev1beta1.DynaKube
+
+		err := dynatracev1beta1.AddToScheme(resources.GetScheme())
+		require.NoError(t, err)
+
+		err = resources.Get(ctx, "dynakube", "dynatrace", &dk)
+		require.NoError(t, err)
+
+		dk.Status.UpdatedTimestamp = metav1.Now()
+		dk.Spec.OneAgent.CloudNativeFullStack.Version = semanticVersion.String()
+		err = resources.Update(ctx, &dk)
+		require.NoError(t, err)
+
+		return ctx
+	}
+}
+
+func restartCSIDriver(ctx context.Context, t *testing.T, environmentConfig *envconf.Config) context.Context {
+	resources := environmentConfig.Client().Resources()
+	var csiDaemonSet appsv1.DaemonSet
+
+	err := csi.ForEachPod(ctx, resources, func(podItem corev1.Pod) {
+		require.NoError(t, resources.Delete(ctx, &podItem))
+	})
+
+	require.NoError(t, resources.Get(ctx, "dynatrace-oneagent-csi-driver", "dynatrace", &csiDaemonSet))
+	require.NoError(t, wait.For(conditions.New(resources).ResourceMatch(&csiDaemonSet, func(object k8s.Object) bool {
+		daemonSet, isDaemonSet := object.(*appsv1.DaemonSet)
+		return isDaemonSet && daemonSet.Status.NumberReady == daemonSet.Status.NumberAvailable
+	}), wait.WithTimeout(10*time.Minute)))
+
+	require.NoError(t, err)
+	time.Sleep(time.Minute)
+
+	return ctx
+}
+
+func checkVersionInSampleApp(semanticVersion version.SemanticVersion) features.Func {
+	return func(ctx context.Context, t *testing.T, environmentConfig *envconf.Config) context.Context {
+		time.Sleep(time.Minute)
+		resources := environmentConfig.Client().Resources()
+		pods := sampleapps.Get(ctx, t, resources)
+
+		for _, podItem := range pods.Items {
+			require.NotNil(t, podItem)
+			require.NotNil(t, podItem.Spec)
+			require.NotNil(t, podItem.Spec.InitContainers)
+			initContainer := podItem.Spec.InitContainers[0]
+			if initContainer.Name == "install-oneagent" {
+				require.NotNil(t, initContainer.Env)
+				require.True(t, kubeobjects.EnvVarIsIn(initContainer.Env, agentVersion))
+				for _, env := range initContainer.Env {
+					if env.Name == agentVersion {
+						require.NotNil(t, env.Value)
+						assert.Equal(t, semanticVersion.String(), env.Value)
+					}
+				}
+			}
+		}
```

Comment:
well, this is deeep :D 

probably could be refactored to be more readable, by moving things to helper functions 😉